synapse/api/filtering.py

```diff
@@ -351,13 +354,13 @@ def __init__(self, hs: "HomeServer", filter_json: JsonDict):
             self.not_rel_types = filter_json.get("org.matrix.msc3874.not_rel_types", [])
 
     def filters_all_types(self) -> bool:
-        return "*" in self.not_types
+        return self.types == [] or "*" in self.not_types
 
     def filters_all_senders(self) -> bool:
-        return "*" in self.not_senders
+        return self.senders == [] or "*" in self.not_senders
 
     def filters_all_rooms(self) -> bool:
-        return "*" in self.not_rooms
+        return self.rooms == [] or "*" in self.not_rooms
```

Comment:
They default to `None` in `__init__`, so I guess we rely on the fact that the JSON validator is putting `[]` as a default value or something like that ?

To avoid errors I think I would change `__init__` to default to `[]` instead of `None`, it looks cleaner to me and on par with `not_rooms`.

This comment also applies to `types` and `senders`.

Reply:
They are not equivalent: `None` means that the filter did not include that key, while `[]` means the filter explicitly passed a value of `[]`.

Reply:
Indeed, I just noticed the comment regarding absent in the spec.